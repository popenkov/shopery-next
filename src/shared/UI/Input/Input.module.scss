@import '@/shared/styles/mixins/text-styles.scss';
@import '@/shared/styles/mixins/box.scss';

.inputWrapper {
  position: relative;
  display: block;
  width: 100%;
}

.input {
  @include search-field;

  width: 100%;
  padding: 14px 16px;
  border: 1px solid rgb(230 230 230);
  border-radius: 6px;
  outline: none;
  transition: all var(--transition);

  &::placeholder {
    color: var(--gray-500);
  }

  &:focus {
    border-color: var(--primary);
  }

  &.search {
    position: relative;
    padding: 14px 16px 14px 44px;
    border: 1px solid rgb(230 230 230);
    border-radius: 6px;
    outline: none;
  }
}

.searchIcon {
  @include box(20px);

  position: absolute;
  top: 12px;
  left: 15px;
}
