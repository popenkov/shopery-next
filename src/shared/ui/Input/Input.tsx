'use client';

import { ChangeEvent, InputHTMLAttributes, forwardRef, memo, useId } from 'react';
import cn from 'classnames';

import AlertIcon from '@public/icons/icon__alert.svg';
import SearchIcon from '@public/icons/icon__search.svg';
import SuccessIcon from '@public/icons/icon__success.svg';

import { Text } from '../Text';

import cls from './Input.module.scss';

type HTMLInputProps = Omit<
  InputHTMLAttributes<HTMLInputElement>,
  'value' | 'onChange' | 'readOnly'
>;

interface Props extends HTMLInputProps {
  className?: string;
  label?: string;
  errorText?: string;
  hasValidation?: boolean;
  value?: string;
  onChange?: (evt: ChangeEvent<HTMLInputElement>) => void;
}

export const Input = memo(
  forwardRef<HTMLInputElement, Props>((props, ref) => {
    const {
      className,
      value,
      label,
      type = 'text',
      errorText,
      hasValidation = true,
      onChange,
      ...otherProps
    } = props;

    //   // todo
    //   // const handleInputChange = (evt: ChangeEvent<HTMLInputElement>) => {
    //   //   // if (type === 'email') {
    //   //   //   setIsInvalid(!checkEmailValidation(value?.toString() || ''));
    //   //   // }

    //   //   onChange?.(evt.target.value);
    //   // };

    // todo
    const handleLoupeClick = () => {
      // ref?.current?.focus();
    };

    const uniqueId = useId();

    return (
      <div className={cls.inputContainer}>
        {label && (
          <label htmlFor={uniqueId} className={cls.inputLabel}>
            {label}
          </label>
        )}
        <div className={cn(cls.inputWrapper, className)}>
          <input
            id={uniqueId}
            ref={ref}
            type={type}
            // value={value}
            onChange={(evt) => onChange?.(evt)}
            className={cn(cls.input, {
              [cls.search]: type === 'search',
              [cls.error]: errorText,
              [cls.success]: value && !errorText,
            })}
            {...otherProps}
          />
          {type === 'search' && (
            <SearchIcon
              className={cn(cls.inputIcon, cls.inputIconSearch)}
              onClick={handleLoupeClick}
            />
          )}
          {/* todo */}
          {hasValidation && (
            <>
              {errorText && <AlertIcon className={cn(cls.inputIcon, cls.inputIconError)} />}
              {/* <WarningIcon className={cn(cls.inputIcon, cls.inputIconWarning)} /> */}
              {value && !errorText && (
                <SuccessIcon className={cn(cls.inputIcon, cls.inputIconSuccess)} />
              )}
            </>
          )}

          {errorText && (
            <Text variant="body_s" weight="medium" className={cls.errorMessage} as="span">
              {errorText}
            </Text>
          )}
        </div>
        {/* todo */}
        {/* {isInvalid && hasValidation && <span>error</span>} */}
      </div>
    );
  }),
);

Input.displayName = 'Input';
