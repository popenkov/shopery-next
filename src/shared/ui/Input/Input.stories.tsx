import type { Meta, StoryObj } from '@storybook/react';

import { Input } from './Input';
import { userEvent, within } from '@storybook/test';

const meta = {
    title: 'shared/Input',
    component: Input,
    parameters: {
        layout: 'centered',
    },
    tags: ['autodocs'],
} satisfies Meta<typeof Input>;

export default meta;
type Story = StoryObj<typeof meta>;

export const InputComponent: Story = {
    decorators: [
        (Story) => (
            <div style={{ maxWidth: '400px' }}>
                <Story />
            </div>
        ),
    ],
    args: { placeholder: 'name' },
};

export const InputFilledComponent: Story = {
    decorators: [
        (Story) => (
            <div style={{ maxWidth: '400px' }}>
                <Story />
            </div>
        ),
    ],
    args: {
        placeholder: 'name',
    },
    play: async ({ canvasElement }) => {
        const canvas = within(canvasElement);

        const emailInput = canvas.getByPlaceholderText('name');

        await userEvent.type(emailInput, 'example-email@email.com', {
            delay: 100,
        });
    },
};
