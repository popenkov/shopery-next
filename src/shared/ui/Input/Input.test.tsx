import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import { Input } from './Input';

describe('Input component', () => {
  it('renders correctly', () => {
    const { container } = render(<Input />);
    expect(container).toMatchSnapshot();
  });

  it('renders label correctly', () => {
    const label = 'Input label';
    const { getByTestId } = render(<Input label={label} />);
    const labelElement = getByTestId('Input.label');
    expect(labelElement).toBeInTheDocument();
  });

  it('renders input field correctly', () => {
    const { getByTestId } = render(<Input />);
    const inputField = getByTestId('Input.input');
    expect(inputField).toBeInTheDocument();
  });

  it('calls onChange handler when input value changes', () => {
    const onChange = jest.fn();
    const { getByTestId } = render(<Input onChange={onChange} />);
    const inputField = getByTestId('Input.input');
    fireEvent.change(inputField, { target: { value: 'New value' } });
    expect(onChange).toHaveBeenCalledTimes(1);
  });

  it('renders error message correctly', () => {
    const errorText = 'Error message';
    const { container } = render(<Input errorText={errorText} />);
    const errorMessage = container.querySelector('.errorMessage');
    expect(errorMessage).toBeInTheDocument();
  });

  it('renders success icon correctly', () => {
    const { container } = render(<Input value="Valid value" />);
    const successIcon = container.querySelector('.inputIconSuccess');
    expect(successIcon).toBeInTheDocument();
  });

  it('renders search icon correctly', () => {
    const { container } = render(<Input type="search" />);
    const searchIcon = container.querySelector('.inputIconSearch');
    expect(searchIcon).toBeInTheDocument();
  });

  //   it('calls handleLoupeClick when search icon is clicked', () => {
  //     const handleLoupeClick = jest.fn();
  //     const { container } = render(<Input type="search" />);
  //     const searchIcon = container.querySelector('.inputIconSearch');
  //     fireEvent.click(searchIcon!);
  //     expect(handleLoupeClick).toHaveBeenCalledTimes(1);
  //   });
});
