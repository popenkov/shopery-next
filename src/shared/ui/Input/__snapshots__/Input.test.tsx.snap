// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component renders correctly 1`] = `
<div>
  <div
    class="inputContainer"
  >
    <div
      class="inputWrapper"
    >
      <input
        class="input"
        data-testid="Input.input"
        id=":r0:"
        type="text"
      />
    </div>
  </div>
</div>
`;
